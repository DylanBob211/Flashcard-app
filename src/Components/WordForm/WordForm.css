input, button {
    appearance: none;
    -webkit-appearance: none;
    border: none;
}

.wordForm_container {
    display: flex;
    flex-direction: row;
    justify-content: center;
}

.wordForm_input {
    margin-bottom: .4em;
    border: solid rgba(0, 0, 0, 0.3) 1px;
    border-radius: 0 0 5px 5px;
    padding: .5em;
}

.wordForm_btn {
    position: relative;
    width: 2em;
    height: 2em;
    margin-left: .5em;
    margin-bottom: .4em;
    background-color: var(--primary-color);
    opacity: 1;
    transition: .3s all;
    border : solid 3px rgba(145, 243, 197, 0.5);
    border-radius: 5px;

}
.wordForm_btn:hover{
    transform : scale(1.3);
    border-radius: 7px;

}

.wordForm_btn::after{
    content: '';
    position: absolute;
    background-color: #fff;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    z-index: 1;
    height: .25em;
    width: 1em;
    border-radius: 3px;
}

.wordForm_btn::before{
    content: '';
    position: absolute;
    background-color: #fff;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    z-index: 2;
    height: 1em;
    width: .25em;
    border-radius: 3px;
}



