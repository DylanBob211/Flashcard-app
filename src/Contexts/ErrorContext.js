import React, { useState, useContext } from 'react';

const ErrorContext = React.createContext(['', () => {}]);

export const useErrorContext = () => useContext(ErrorContext);

const ErrorProvider = ({ children }) => {
  const [error, setError] = useState('');
  return (
    <ErrorContext.Provider
      value={[
        error,
        setError,
      ]}
    >
      {children}
    </ErrorContext.Provider>
  );
};

export default ErrorProvider;
