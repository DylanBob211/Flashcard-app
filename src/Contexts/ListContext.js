import React, { useContext } from 'react';
import initialState from '../Components/initialState';
import useLists from '../Hooks/useLists';

const ListContext = React.createContext([initialState, () => {}]);

export const useListContext = () => useContext(ListContext);

const ListProvider = ({ children }) => {
  const { lists, setLists, addNewList } = useLists();
  return (
    <ListContext.Provider value={{ lists, setLists, addNewList }}>
      {children}
    </ListContext.Provider>
  );
};

export default ListProvider;
