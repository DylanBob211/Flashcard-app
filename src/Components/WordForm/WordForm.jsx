import React, { useState } from 'react';
import PropTypes from 'prop-types';
import './WordForm.css';
import { useErrorContext } from '../../Contexts/ErrorContext';
import PlusIcon from '../Icons/PlusIcon';


const WordForm = ({ addWord }) => {
  const [, handleError] = useErrorContext();
  const [newWord, setNewWord] = useState('');

  const createNewWord = (e) => {
    e.preventDefault();
    if (!newWord) {
      handleError('Add a word first');
    }
    addWord(newWord, 'it')
      .catch(error => console.error(error.message));

    setNewWord('');
  };

  const handleChange = (e) => {
    e.persist();
    handleError('');
    setNewWord(e.target.value);
  };

  return (
    <form
      className="wordForm_container hide-submit"
      onSubmit={e => createNewWord(e)}
    >
      <input
        className="wordForm_input"
        type="text"
        onChange={e => handleChange(e)}
        value={newWord}
        placeholder="Add word here..."
      />
      <label data-test="addWordButtonTest">
        <input type="submit" />
        <PlusIcon
          disabled={!addWord}
        />
      </label>
    </form>
  );
};

WordForm.propTypes = {
  addWord: PropTypes.func,
};

WordForm.defaultProps = {
  addWord: null,
};

export default WordForm;
