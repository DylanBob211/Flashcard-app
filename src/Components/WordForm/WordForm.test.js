import React from 'react';
import { mount, shallow } from 'enzyme';
import * as ErrorContext from '../../Contexts/ErrorContext';
import WordForm from './WordForm';

describe('WordForm Component', () => {
  const handleError = jest.fn();
  const useErrorContext = jest.spyOn(ErrorContext, 'useErrorContext');
  useErrorContext.mockImplementation(() => [null, handleError]);

  let mountedWordForm;

  const wordForm = (props, isShallow = false) => {
    const jsx = (<WordForm {...props} />);
    if (isShallow) return shallow(jsx);
    return mount(jsx);
  };

  const props = {
    addWord: null,
  };

  beforeEach(() => {
    mountedWordForm = undefined;
    jest.clearAllMocks();
  });

  it('Matches to snapshot', () => {
    const component = wordForm(props, true);
    expect(component).toMatchSnapshot();
  });

  describe('always', () => {
    beforeEach(() => {
      mountedWordForm = wordForm(props);
    });

    it('renders a form container', () => {
      const formContainer = mountedWordForm.find('.wordForm_container');
      expect(formContainer.exists()).toEqual(true);
    });

    it('renders an input', () => {
      const addWordInput = mountedWordForm.find('input.wordForm_input');
      expect(addWordInput.exists()).toEqual(true);
    });

    it('renders a PlusIcon button', () => {
      const plusIcon = mountedWordForm.find('PlusIcon');
      expect(plusIcon.exists()).toEqual(true);
    });

    it('disables the button if no addWord prop is passed', () => {
      const plusIcon = mountedWordForm.find('PlusIcon');
      expect(plusIcon.props().disabled).toEqual(true);
    });
  });

  describe('when `addWord` is passed', () => {
    beforeEach(() => {
      props.addWord = jest.fn();
      mountedWordForm = wordForm(props);
    });

    it('enables the the button', () => {
      const plusIcon = mountedWordForm.find('PlusIcon');
      expect(plusIcon.props().disabled).toEqual(false);
    });

    it('calls handleError with "Add a word first" if form is submitted', () => {
      const plusIcon = mountedWordForm.find('PlusIcon');
      plusIcon.simulate('submit');
      expect(handleError).toHaveBeenCalledWith('Add a word first');
    });

    describe('when inputting', () => {
      const testInput = 'Hello Input';
      beforeEach(() => {
        const input = mountedWordForm.find('.wordForm_input');
        input.simulate('change', { target: { value: testInput } });
      });

      it('calls handleError with empty string inputting', () => {
        expect(handleError).toHaveBeenCalledWith('');
      });

      it('changes the value of the input', () => {
        const input = mountedWordForm.find('.wordForm_input');
        expect(input.props().value).toEqual(testInput);
      });

      describe('when submitting', () => {
        beforeEach(() => {
          const plusIcon = mountedWordForm.find('PlusIcon');
          plusIcon.simulate('submit');
        });

        it('calls addWord with the input inserted', () => {
          expect(props.addWord).toHaveBeenCalledWith(testInput);
        });

        it('empties the input', () => {
          const input = mountedWordForm.find('.wordForm_input');
          expect(input.props().value).toEqual('');
        });
      });
    });
  });
});
