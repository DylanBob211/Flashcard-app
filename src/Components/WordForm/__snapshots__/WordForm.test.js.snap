// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordForm Component Matches to snapshot 1`] = `
<form
  className="wordForm_container hide-submit"
  onSubmit={[Function]}
>
  <input
    className="wordForm_input"
    onChange={[Function]}
    placeholder="Add word here..."
    type="text"
    value=""
  />
  <label
    data-test="addWordButtonTest"
  >
    <input
      type="submit"
    />
    <PlusIcon
      className=""
      disabled={true}
    />
  </label>
</form>
`;
